### ISSUE TITLE
How to use ContrastiveLoss

### ISSUE THREAD

**Issue by icsd13152** [KevinMusgrave/pytorch-metric-learning#549]
Hi,

I have a Siamese NN architercture that holds a Bert Transformer for each of the siblings Sub-networks.
So I have sentences pairs and I want to encode each of the sentence in order to get their Embeddings.

def forward(set1, set2):
      embeddings1 = BERT(set1)  #here we have batch_size x 768 after avg pooling
      embeddings2 = BERT(set2)  #here we have batch_size x 768 after avg pooling
      return embeddings1, embeddings2

I am wondering how to pass those 2 Embeddings Tensors to ContrastiveLoss? Should I concatenate them in 0 dimension in order to have 2 x batch_size x 768 -----> for example if batch size equals to 16 then after concat I will have 32x768. 
After that I must repeat the tensor of Labels. I suppose.....
Labels.repeat(2) #32x768

I can not understand how the loss between Embeddings is calculated. It calculates first the distance for every possible pair in 32 batch_size, based on labels?

Could you please provide me an example?

Thanks in advanced

**Comment by KevinMusgrave** [comment]
> I am wondering how to pass those 2 Embeddings Tensors to ContrastiveLoss? Should I concatenate them in 0 dimension in order to have 2 x batch_size x 768 -----> for example if batch size equals to 16 then after concat I will have 32x768. After that I must repeat the tensor of Labels. I suppose..... Labels.repeat(2) #32x768

Yes you should concatenate the embeddings and concatenate the labels.

> I can not understand how the loss between Embeddings is calculated. It calculates first the distance for every possible pair in 32 batch_size, based on labels?

Yes it computes the loss based on all positive pairs and negative pairs in the batch of 32.

> Could you please provide me an example?

In the following, I assume `embeddings1[i]` is supposed to form a positive pair with `embeddings2[i]`:

```python
# your training for-loop
optimizer.zero_grad()
embeddings1, embeddings2 = model(data)
labels = torch.arange(embeddings1.size(0))

embeddings = torch.cat([embeddings1, embeddings2 ], dim=0)
labels = torch.cat([labels, labels], dim=0)

loss = loss_func(embeddings, labels)
loss.backward()
optimizer.step()
```

**Comment by icsd13152** [comment]
Ok. Perfect!  Thank you very much for your response. 

I was using the same way you provided to me, but while the loss is decreased the cosine similarity between embeddings in never decreased for negative pairs and incresed for positive pairs. Below a piece of code. Maybe something goes wrong with the weight of architecture.

`class myModelEmbeddings(nn.Module):
  def __init__(self,bert_emb_layer,startLayer,endLayer,bertModel,groupLayersMode = (False,False)):#(True,True)-> Grouping and Summing | #(True,False)-> Grouping and Concat
      super(myModelEmbeddings, self).__init__()
      self.bert_emb_layer = bert_emb_layer
      self.startLayer = startLayer
      self.endLayer = endLayer

      self.groupLayersMode = groupLayersMode
      self.bertModel = bertModel
      self.bertModel.train()

      inputFeatures = 0
      if self.groupLayersMode == (True,False):
        inputFeatures = (endLayer - startLayer)*768 
      elif self.groupLayersMode == (True,True):
        inputFeatures = 768
      else:
        inputFeatures = 768
      self.dropout = nn.Dropout(0.5)
      self.FC = nn.Sequential(
       
        nn.Linear(in_features = inputFeatures,out_features = 32),
      )

  def getSpecificLayerOfBERT(self,bertOutputs):
      hidden_states = bertOutputs[2][1:] 
      layerOutput = hidden_states[self.bert_emb_layer] # get specific Layer (from 0 to 11) for all tuples (batch_size, sequence_length, hidden_size)
      
      return  layerOutput
  
  def concatSpecificLayersOfBERT(self,bertOutputs):
      hidden_states = bertOutputs[2][0:] 
      concatEmbeddingLayers = torch.cat([hidden_states[i] for i in range(self.startLayer,self.endLayer)], dim=-1)
      
      return concatEmbeddingLayers
  def getCLSEmbeddings(self,bertOutputs ):
      embeddings = bertOutputs[1]
      return embeddings
  def sumSpecificLayersOfBERT(self,bertOutputs):
      #Number of layers: 13   (initial embeddings + 12 BERT layers) - So we need [2][1:] 1 and onwards
      hidden_states = bertOutputs[2][0:]
      # `hidden_states` is a Python list.
      
      sumEmbeddingLayers = torch.stack(hidden_states[self.startLayer:self.endLayer]).sum(0)
      del hidden_states

      return sumEmbeddingLayers
  def pooling(self,token_embeddings, mask, strategy='avg'):
      if strategy == 'max':
         input_mask_expanded = mask.unsqueeze(-1).expand(token_embeddings.size()).float()
         token_embeddings[input_mask_expanded == 0] = -1e9  # Set padding tokens to large negative value
         max_setence_embeddings = torch.max(token_embeddings, 1)[0]
         return max_setence_embeddings
      elif strategy == 'avg':
         in_mask = mask.unsqueeze(-1).expand(token_embeddings.size()).float()
         # perform mean-pooling but exclude padding tokens (specified by in_mask)
         avg_setence_embeddings = torch.sum(token_embeddings * in_mask, 1) / torch.clamp(in_mask.sum(1), min=1e-9)
         return avg_setence_embeddings
      elif strategy == 'sum':
        sum_setence_embeddings = torch.sum(token_embeddings[0:len(token_embeddings)],1)
        return sum_setence_embeddings

  def forwardOnce(self, sent_id, mask):
      outputs =  self.bertModel(sent_id, attention_mask=mask)

      if self.groupLayersMode == (True,False):
        embeddings = self.concatSpecificLayersOfBERT(outputs)
        return  embeddings #, self.FC(embeddings)
      elif self.groupLayersMode == (True,True):
        embeddings = self.sumSpecificLayersOfBERT(outputs)
        return embeddings #, self.FC(embeddings)
      else:
        embeddings = self.getSpecificLayerOfBERT(outputs)

        return embeddings #, self.FC(embeddings)

  def forward(self, sent_id1, mask1,sent_id2, mask2,b_labels):

      # forward pass of input 1
      output1 = self.forwardOnce(sent_id1, mask1)
      # forward pass of input 2
      output2  = self.forwardOnce(sent_id2, mask2)

      avgoutput1 = self.pooling(output1,mask1,'avg')
      avgoutput2 = self.pooling(output2,mask2,'avg')

     
      FC11 = self.FC(avgoutput1)
      FC22 = self.FC(avgoutput2)

      return FC11, FC22,output`

and in train loop I have the below

`def train():
     FC11,FC22,cos = model(sentences1,sentences2,b_labels)
     embeddings = torch.cat([ FC11,FC22], dim=0)  
     loss = criterion1(embeddings ,b_labels.repeat(2))
#of cource here loss.backward etc etc`

Is that way correct? Am I doing something wrong? I am wondering if the weight are shared between the 2 berts. it's one bert and I am using it for feuture extraction twice.

**Comment by icsd13152** [comment]
Sorry for the spam. I am a little bit confused about how the loss is calculated.

If I have the below example how the loss function computes the distance between embeddings and as a result a loss value?

embeddings1[0] -- embeddings2[0] --> Positive pair (label=1)
embeddings1[1] -- embeddings2[1] --> Positive pair (label=1)
embeddings1[2] -- embeddings2[2] --> Negative pair (label=0)

As we can observe we have batch_size=3.

After concat I will have the below with batch_size=6:

embeddings1[0]  --> (label=1)
embeddings1[1]	--> (label=1)
embeddings1[2]	--> (label=0)
embeddings2[0]	--> (label=1)
embeddings2[1]	--> (label=1)
embeddings2[2]	--> (label=0)

The first calculation of the loss is the distance between vectors. So the Contrastive Loss function will calculate the distances in the same way as below? (one by one comparison)

|embeddings1[0] - embeddings1[1]|
|embeddings1[0] - embeddings1[2]|
|embeddings1[0] - embeddings2[0]|
|embeddings1[0] - embeddings2[1]|
|embeddings1[0] - embeddings2[2]|

and it continue as above for every possible pair in batch?

Or it will calcualte the distances as below? (based on labels)

|embeddings1[0] - embeddings2[0]| -- Positive pair
|embeddings1[1] - embeddings2[1]| -- Positive pair
|embeddings1[2] - embeddings2[2]| -- Negative pair

To do that I suppose there is a piece of code that divides the length of Embeddings Tensor.

Could you please provide me a break down of calculations based on the above example?

Thanks in advanced

**Comment by KevinMusgrave** [comment]
> Is that way correct? Am I doing something wrong? I am wondering if the weight are shared between the 2 berts. it's one bert and I am using it for feuture extraction twice.

Looks ok to me.

> The first calculation of the loss is the distance between vectors. So the Contrastive Loss function will calculate the distances in the same way as below? (one by one comparison)
> 
> |embeddings1[0] - embeddings1[1]|
> |embeddings1[0] - embeddings1[2]|
> |embeddings1[0] - embeddings2[0]|
> |embeddings1[0] - embeddings2[1]|
> |embeddings1[0] - embeddings2[2]|

That's correct


> 
> Or it will calcualte the distances as below? (based on labels)
> 
> |embeddings1[0] - embeddings2[0]| -- Positive pair 
> |embeddings1[1] - embeddings2[1]| -- Positive pair 
> |embeddings1[2] - embeddings2[2]| -- Negative pair

If you want vectors from `embeddings1` to only be compared to vectors from `embeddings2` then you can use the `ref_emb` argument:

```python
embeddings1, embeddings2 = model(data)
loss = loss_func(embeddings1, labels, ref_emb=embeddings2, ref_labels=labels)
```

This way, all pairs will be of the form `(embeddings1[i], embeddings2[j])`.

**Comment by icsd13152** [comment]
>Using ref_emb with your example, the positive pairs will be:

embeddings1[0], embeddings2[0]
embeddings1[0], embeddings2[1]
embeddings1[1], embeddings2[0]
embeddings1[1], embeddings2[1]
embeddings1[2], embeddings2[2]

I see... But If I have the indices of pairs strictly defined?

For example 
I want to compute loss for 
embeddings1[0] - embeddings2[0]
.
.
.
embeddings1[i] - embeddings2[i] (index of embeddings1 is the same with the index of embeddings2)
I do not want to compare embeddings1[i] with embeddings[i+n] (where n is an integer )
Something like below

embeddings1[0] - embeddings2[0] --> Positive 
embeddings1[1] - embeddings2[1] --> Positive
embeddings1[2] - embeddings2[2] --> Negative

I want to keep the indices of embeddings1 same with embeddings2. How cat I do it in this way?
Below an example in order to understand further why I am asking that.
Let's say that the Embeddings are produced based on sentences that have been written from authors.

Embeddings1[0] - Embeddings2[0] --> Same Author, but different Sentences. (Positive Label) AuthorID=1 That is correct
Embeddings1[0] - Embeddings2[1] --> Negative Label, but maybe here we have Same Authors. So the Actual label would be Positive. That is not-correct.

In other words I want to keep the predefined indices of pairs and not compare-contrast all the possitives and all the negatives because maybe some negative Labels will be actually positive.

What about label indices?

Thanks in advanced

**Comment by KevinMusgrave** [comment]
To use predefined pairs, use the `indices_tuple` argument.

```python
# using your example
p = torch.tensor([0, 1])
n = torch.tensor([2])
indices_tuple = (p,p,n,n)
loss = loss_func(embeddings1, labels, indices_tuple, ref_emb=embeddings2, ref_labels=labels)
```

The first 2 tensors of `indices_tuple` form all the positive pairs. In the example above, these are `[0, 1]` and `[0, 1]`, which means that the first positive pair is `emb1[0], emb2[0]`, and the second positive pair is `emb1[1], emb2[1]`. 

The other 2 tensors form all the negative pairs. In the example above, these are `[2]` and `[2]`, which means there is one negative pair: `emb1[2], emb2[2]`.

**Comment by icsd13152** [comment]
> To use predefined pairs, use the `indices_tuple` argument.
> 
> ```python
> # using your example
> p = torch.tensor([0, 1])
> n = torch.tensor([2])
> indices_tuple = (p,p,n,n)
> loss = loss_func(embeddings1, labels, indices_tuple, ref_emb=embeddings2, ref_labels=labels)
> ```
> 
> The first 2 tensors of `indices_tuple` form all the positive pairs. In the example above, the first two tensors are `[0, 1]` and `[0, 1]`, which means that the first positive pair is `emb1[0], emb2[0]`, and the second positive pair is `emb1[1], emb2[1]`.
> 
> The other 2 tensors form all the negative pairs. In the example above, these tensors are `[2]` and `[2]`, which means there is one negative pair: `emb1[2], emb2[2]`.

Ohhh GREAT! Thank you very much. I think this will solve my issues. 
Thanks a lot Kevin. 

Best Regards

**Comment by icsd13152** [comment]
Hi Kevin,

I tried the solution wyou provided to me, but as per documentation if i set margin_pos=1 and margin_neg=0 (I am using Cosinine Similarity) the loss is never decreased an it's almost 0.95.  (train and val loss)

With pos_margin=0 and neg_margin=0 the loss is decreased linearly to almost 0.03. (train and val loss)

Why this behavor ? Am I doing something wrong or is something I miss in this specific loss calculation?
Could you please clariffy further how this loss is calculated?

Thanks

**Comment by KevinMusgrave** [comment]
> Why this behavor ? Am I doing something wrong or is something I miss in this specific loss calculation? Could you please clariffy further how this loss is calculated?

The loss for positive pairs is `max(0, pos_margin - sim(x, y)` and the loss for negative pairs is `max(0, sim(x, y) - neg_margin`, where `sim(x, y)` is the cosine similarty between two embeddings.

> Can I set negative values for neg_margin? For example if I want to get a simialrity less than this margin , Can I set the margin=-0.5?

Yes you can try a negative value for `neg_margin`.

**Comment by icsd13152** [comment]
Thank you for your response.

It seems to me that is happening the opposite from what I want. 
For positive pairs the loss pushes the vectors as far as possible.
For negative pairs pushes them closer.

I have tried different values for Learning rate and different values for margins. 
I have also tried different datasets. Always same issues. 

My labels are always 1 for positives and 0 for negatives. 

From your experience could you please give me a tip of what else maybe cause those issues?

**Comment by KevinMusgrave** [comment]
It should be ok if you're following the `indices_tuple` method I described above: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/549#issuecomment-1312570080

You can also try the normalized L2 distance, i.e. don't pass in a distance function, and set `pos_margin = 0` and `neg_margin = 1`. 

I'm not sure why the positive pairs are being pushed apart and the negative pairs are being pushed together.

**Comment by icsd13152** [comment]
I have tried it with L2 distance. 
Due to the fact that I am using BERT as siemese network for embeddings extractor I can not increase the batch size. I am traing the NN with batch =16
Maybe larger batch size could be helpfull. Please tell me your opinion.

**Comment by KevinMusgrave** [comment]
In your setup, your batch size is equal to the number of pairs. 16 pairs is quite low, so a larger batch size would help. If you can't fit it memory you could try [gradient accumulation](https://kozodoi.me/python/deep%20learning/pytorch/tutorial/2021/02/19/gradient-accumulation.html).

**Comment by icsd13152** [comment]
> In your setup, your batch size is equal to the number of pairs. 16 pairs is quite low, so a larger batch size would help. If you can't fit it in memory you could try [gradient accumulation](https://kozodoi.me/python/deep%20learning/pytorch/tutorial/2021/02/19/gradient-accumulation.html).

I will try it and let you know. Thank you very much

**Comment by icsd13152** [comment]
I tried with batch size 64 without accumulation and pos_margin=1 and neg_margin=-1. I didn't get any OOM. 
The results were too good for example cosine similarities for negative pairs pushed towards negative values and the opposite for positive pairs. The problem with loss still exists. The loss never decreased and the loss values were between 0.99 and 1.3. 
Why that?

**Comment by KevinMusgrave** [comment]
Because the loss is the average of non-zero losses. So many of the pairs may have 0 loss, and they don't contribute to the final value: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/504#issuecomment-1201083749

**Comment by icsd13152** [comment]
I see... the model can seperate negative pairs and positive pairs but the loss may be to high. I think that is a problem, because when I want to report in a paper the trade-off of losses over the epochs, it will be reported with too high error.

**Comment by KevinMusgrave** [comment]
If you want it to take the average of all losses and no just the non-zero losses, you can use MeanReducer:

```python
from pytorch_metric_learning.reducers import MeanReducer

loss_fn = ContrastiveLoss(reducer=MeanReducer())
```

**Comment by icsd13152** [comment]
Same results with MeanReducer. I inspected all the batches in validation set and seems that the model correctly decrease/increase the similarities. Expect form some pairs that model fails, but this is acceptable. Why those too high values in loss per batch? I call the .item() function to get the loss per batch. I suppose that it's correct.

**Comment by KevinMusgrave** [comment]
It's hard to say why specific pairs have high loss. It could just be that those are difficult pairs to learn.

If you think the model is overfitting on your training set, then you need a larger dataset, or you need data augmentations. I'm not familiar with data augmentations for NLP though.

**Comment by KevinMusgrave** [comment]
`SelfSupervisedLoss` has been added to v2.0.0. If you have two "views" of data, you don't need to make labels anymore:

```python
from pytorch_metric_learning.losses import SelfSupervisedLoss
loss_func = SelfSupervisedLoss(NTXentLoss())
embeddings = model(data)
augmented = model(augmented_data)
loss = loss_func(embeddings, augmented)
```

**Comment by pheersaquoff** [comment]
@KevinMusgrave hello, I'm also trying to use contrsative loss with BaseTrainer  and get an error, could you help please
```
loss_funcs = {
        'metric_loss': losses.ContrastiveLoss()
    }
    optimizers = {
        'trunk_optimizer': torch.optim.Adam(models['trunk'].parameters(), lr=params['trunk_lr'],
                                                      weight_decay=params['trunk_decay']),
        'embedder_optimizer': torch.optim.Adam(models['embedder'].parameters(), lr=params['emb_lr'],
                                                         weight_decay=params['emb_decay'])
    }
    lr_schedulers = {
        'trunk_scheduler_by_plateau': torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer=optimizers['trunk_optimizer'], patience=10),
        'embedder_scheduler_by_plateau':   torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer=optimizers['embedder_optimizer'], patience=10) 
    }

    batch_size = params['support_batch'] if test else params['batch']
    dataset_dict = {'val': val_dataset}

    if test: #where to put final weights/models
        base_folder = f"../exp/classifier/{params['exp_name']}/support/"
    else:
        base_folder = f"../exp/classifier/{params['exp_name']}/"
    model_folder = base_folder + 'saved_models'
    record_keeper, _, _ = logging_presets.get_record_keeper(
        base_folder + 'logs', base_folder + 'tensorboard'
    )

    hooks = logging_presets.get_hook_container(record_keeper)
    tester = testers.GlobalEmbeddingSpaceTester(
        end_of_testing_hook=hooks.end_of_testing_hook,
        dataloader_num_workers=params['workers'],
        accuracy_calculator=AccuracyCalculator(k='max_bin_count'),
    )
    end_of_epoch_hook = hooks.end_of_epoch_hook(
        tester, dataset_dict, model_folder, test_interval=5, patience=30
    )
    m = 4 
    trainer = trainers.BaseTrainer(
        models,
        optimizers,
        batch_size,
        loss_funcs,
        train_dataset,
        #sampler=samplers.MPerClassSampler(train_dataset.targets, m=1, batch_size=batch_size, length_before_new_iter=len(train_dataset)),
        dataloader_num_workers=params['workers'],
        lr_schedulers=lr_schedulers,    
        end_of_iteration_hook=hooks.end_of_iteration_hook,
        end_of_epoch_hook=end_of_epoch_hook,
    )
    return trainer
    
    trainer.train(num_epochs=params['epochs'])
    ```

and I get this 

```
    trainer.train(num_epochs=params['epochs'])
  File "/home/n.firsakov/.conda/envs/ops/lib/python3.10/site-packages/pytorch_metric_learning/trainers/base_trainer.py", line 87, in train
    self.forward_and_backward()
  File "/home/n.firsakov/.conda/envs/ops/lib/python3.10/site-packages/pytorch_metric_learning/trainers/base_trainer.py", line 115, in forward_and_backward
    self.calculate_loss(self.get_batch())
TypeError: BaseTrainer.calculate_loss() takes 1 positional argument but 2 were given
```

**Comment by KevinMusgrave** [comment]
You have to use one of the classes that extends BaseTrainer. Try `trainers.MetricLossOnly`

Sorry for the confusion.

**Comment by pheersaquoff** [comment]
Thank you

**Comment by seanyeo300** [comment]
> > Is that way correct? Am I doing something wrong? I am wondering if the weight are shared between the 2 berts. it's one bert and I am using it for feuture extraction twice.
> 
> Looks ok to me.
> 
> > The first calculation of the loss is the distance between vectors. So the Contrastive Loss function will calculate the distances in the same way as below? (one by one comparison)
> > |embeddings1[0] - embeddings1[1]|
> > |embeddings1[0] - embeddings1[2]|
> > |embeddings1[0] - embeddings2[0]|
> > |embeddings1[0] - embeddings2[1]|
> > |embeddings1[0] - embeddings2[2]|
> 
> That's correct
> 
> > Or it will calcualte the distances as below? (based on labels)
> > |embeddings1[0] - embeddings2[0]| -- Positive pair
> > |embeddings1[1] - embeddings2[1]| -- Positive pair
> > |embeddings1[2] - embeddings2[2]| -- Negative pair
> 
> If you want vectors from `embeddings1` to only be compared to vectors from `embeddings2` then you can use the `ref_emb` argument:
> 
> ```python
> embeddings1, embeddings2 = model(data)
> loss = loss_func(embeddings1, labels, ref_emb=embeddings2, ref_labels=labels)
> ```
> 
> This way, all pairs will be of the form `(embeddings1[i], embeddings2[j])`.
> 
> Using `ref_emb` with your example, the positive pairs will be:
> 
> * embeddings1[0], embeddings2[0]
> * embeddings1[0], embeddings2[1]
> * embeddings1[1], embeddings2[0]
> * embeddings1[1], embeddings2[1]
> * embeddings1[2], embeddings2[2]
> 
> The negative pairs will be:
> 
> * embeddings1[0], embeddings2[2]
> * embeddings1[1], embeddings2[2]
> * embeddings1[2], embeddings2[0]
> * embeddings1[2], embeddings2[1]
> 
> Positive pairs have the same label. In your example, the positive pairs are formed by embeddings that share a label of 1, or that share a label of 0.
> 
> Negative pairs have differing labels. In your example, the negative pairs are formed when `embeddings1[i]` has a label of 1, and `embeddings2[j]` has a label of 0, and vice versa.

Hi Kevin, I am trying to use the ref_emb argument but am getting a TypeError

File:C\...\module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
TypeError: forward() got an unexpected keyword argument 'ref_emb'

My use of the argument is as follows:

loss = loss_func(embeddings1, labels1, ref_emb=embeddings2, ref_labels=labels2)

these are the modules I have imported:

import pytorch_metric_learning
from pytorch_metric_learning import distances, losses, miners, reducers, testers

**Comment by KevinMusgrave** [comment]
@seanyeo300 

You may need to update to a newer version of the library.

The latest version is 2.4.1.

You can print your current version:

```python
import pytorch_metric_learning
print(pytorch_metric_learning.__version__)
```

Install the latest:
```
pip install pytorch-metric-learning==2.4.1
```

**Comment by seanyeo300** [comment]
Thanks for the quick reply, my version was indeed much older. I'll report back if the issue is not resolved, still running some other experiments. Thanks again!
